02_munge/src/munge_usgs.py

```diff
@@ -0,0 +1,79 @@
+    '''
+    process raw data text files into clean csvs, including:
+        dropping unwanted flags
+        converting datetime column to datetime format
+        converting all data columns to numeric type
+        removing metadata columns so that only datetime and data columns remain 
+    '''
+    for raw_datafile in raw_datafiles:
+        # read in raw data as pandas df
+        df = pd.read_csv(raw_datafile, comment='#', sep='\t', lineterminator='\n', low_memory=False)
+        # drop first row which does not contain useful headers or data
+        df.drop(index=0, inplace=True)
+
+        # replace all flagged data we want to remove with NaN
+        flag_cols = [col for col in df.columns if re.match("[0-9]+_[0-9]+_[a-z]+$", col)]
+        for flag_col in flag_cols:
+            flag_data_col = flag_col[:-3]
+            df[flag_data_col] = np.where(df[flag_col].isin(flags_to_drop), np.nan, df[flag_data_col])
+
+        # drop site info and flag columns
+        df = df.drop(['agency_cd', 'site_no', 'tz_cd']+flag_cols, axis=1)
+
+        # convert datetime column to datetime type
+        df['datetime'] = df['datetime'].astype('datetime64')
+
+        # make all other columns numeric
+        cols = df.columns.drop('datetime')
+        df[cols] = df[cols].apply(pd.to_numeric, errors='coerce')
+
+        # (to be added in future) aggregate data to specific timestep?
+        # df = df.groupby([df['datetime'].dt.date])[cols].mean()
+
+        # save pre-processed data
+        data_outfile_csv = os.path.join('.', '02_munge', 'out', os.path.splitext(os.path.basename(raw_datafile))[0]+'.csv')
+        df.to_csv(data_outfile_csv, index=False)
+
+def main():
+    # process raw parameter data into csv
+    raw_params_txt = os.path.join('.', '01_fetch', 'out', 'params.txt')
+    params_outfile_csv = os.path.join('.', '02_munge', 'out', 'params.csv')
+    process_params_to_csv(raw_params_txt, params_outfile_csv)
+
+    # get list of raw data files to process
+    raw_datafiles = [os.path.join('.', '01_fetch', 'out', file) for file in os.listdir(os.path.join('.', '01_fetch', 'out'))]
+    raw_datafiles.remove(os.path.join('.', '01_fetch', 'out', 'params.txt'))
+
+    # determine which data flags we want to drop
+    # e     Value has been edited or estimated by USGS personnel and is write protected
+    # &     Value was computed from affected unit values
+    # E     Value was computed from estimated unit values.
+    # A     Approved for publication -- Processing and review completed.
+    # P     Provisional data subject to revision.
+    # <     The value is known to be less than reported value and is write protected.
+    # >     The value is known to be greater than reported value and is write protected.
+    # 1     Value is write protected without any remark code to be printed
+    # 2     Remark is write protected without any remark code to be printed
+    flags_to_drop = ['e', '&', 'E', 'P', '<', '>', '1', '2']
```

Comment:
This is great, I think what you've written makes sense, but we might want to go back and take a look at when these flags pop up within the timeseries (maybe you've already done this).

Reply:
P (provisional data) was really the main flag I saw popping up a lot, and Jared suggested we not use that data. I noticed a lot of the data files Salme put on Sharepoint included provisional data, and now that data is gone when I fetch from USGS--probably a sign that provisional data is often bad.

Reply:
All of the others sounded kind of like estimations to me too, so I figured if we don't want provisional data, we probably don't want other estimates. The only exception might be for flags 1 and 2...I could try to look into why those might show up.

Are there any flags in particular that you think we might want to use data for?

Reply:
Not that I can think of based on their descriptions